internal/api/promote_subscribers_v1alpha1.go

```diff
@@ -0,0 +1,90 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"connectrpc.com/connect"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kubev1alpha1 "github.com/akuity/kargo/api/v1alpha1"
+	typesv1alpha1 "github.com/akuity/kargo/internal/api/types/v1alpha1"
+	"github.com/akuity/kargo/internal/kargo"
+	"github.com/akuity/kargo/internal/logging"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+	v1alpha1 "github.com/akuity/kargo/pkg/api/v1alpha1"
+)
+
+func (s *server) PromoteSubscribers(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.PromoteSubscribersRequest],
+) (*connect.Response[svcv1alpha1.PromoteSubscribersResponse], error) {
+	if err := validateProjectAndStageNonEmpty(req.Msg.GetProject(), req.Msg.GetStage()); err != nil {
+		return nil, err
+	}
+	if err := s.validateProject(ctx, req.Msg.GetProject()); err != nil {
+		return nil, err
+	}
+	stage, err := getStage(ctx, s.client, req.Msg.GetProject(), req.Msg.GetStage())
+	if err != nil {
+		return nil, err
+	}
+	if err = validateFreightExists(req.Msg.GetFreight(), stage.Status.History); err != nil {
+		return nil, err
+	}
+
+	subscribers, err := s.findStageSubscribers(ctx, stage)
+	if err != nil {
+		return nil, err
+	}
+	if len(subscribers) == 0 {
+		return nil, connect.NewError(connect.CodeNotFound, fmt.Errorf("Stage %q has no subscribers", req.Msg.GetStage()))
+	}
+
+	logger := logging.LoggerFromContext(ctx)
+
+	var promoteErrs []error
+	var createdPromos []*v1alpha1.Promotion
+	for _, subscriber := range subscribers {
+		if err := validateFreightExists(req.Msg.GetFreight(), subscriber.Status.AvailableStates); err != nil {
+			// TODO(JS): currently we create promotions to all of this Stage's subscribers, ignoring
+			// whether or not the freight *also* appears in the availableStates of the subscriber.
+			// Normally, it should always be the case that if it's in our history, it should also
+			// appear in our subscriber's availableStates. For now, just log a warning if we are
+			// promoting something that for some reason, has not yet appeared there.
```

Comment:
Ah. This partially answers my question [here](https://github.com/akuity/kargo/pull/612/files#r1313457336)